### ISSUE TITLE
--strictNullChecks on optional properties too aggressive

### ISSUE THREAD

**Issue by orionlee** [Microsoft/TypeScript#11138]
**TypeScript Version:**  @2.1.0-dev.20160913 (also tested on 0925 nightly)

**Code**
Compile the following code with `--strictNullChecks` flag on

``` ts
class Player {
  private _id? :number = undefined;

  get id() :number {
    if (this._id > 0) {
      return this._id; // tsc reports error TS2322: Type 'number | undefined' is not assignable to type 'number'
    } else {
      throw new Error('Assertion error: .id expected to be a positive number. Actual: ' + this._id);
    }
  }

  // use case requires id be assigned post-constructor
  init(id :number) :void {
    if (id > 0) {
      this._id = id;
    } else {
      throw new TypeError('argument id must be positive number. Actual: ' + id);
    }
  }
}
```

**Expected behavior:**
The code should compile successfully.

**Actual behavior:** 
`tsc` reports error

> tmp.ts(6,14): error TS2322: Type 'number | undefined' is not assignable to type 'number'.
>  Type 'undefined' is not assignable to type 'number'.

**Issue**
The lines in question

```
4:  get id() :number {
5:   if (this._id > 0) {
6:      return this._id;
```

While `_id` is optional typewise, the guard `if(this._id > 0)` guarantees the code returns a defined number. However, `tsc` incorrectly reports it as an error.

**Comment by ahejlsberg** [comment]
@orionlee The behavior you're seeing is by design. Your example works if you first check for `undefined` using a truthy check or an equality comparison, for example

```ts
if (this._id && this._id > 0) ...
```

or

```ts
if (this._id !== undefined && this._id > 0) ...
```

We don't consider the greater-than and less-than operators to be type guards because they have very odd effects with respect to `null` and `undefined`. For example, any comparison involving `undefined` is always `false`, so *both* `x > undefined` and `x < undefined` are false for any non-undefined value of `x`.

**Comment by orionlee** [comment]
@ahejlsberg Thanks for the prompt clarification! 

It would be more productive (to typescript users) if there is have some documentation (or maybe just a well-known informal wiki) to point out similar issues that could stump them.

While I understand your rationale, it is quite tricky. In the case above, the end results for `if (this._id > 0)` and `if (this._id && this._id > 0)` are the same even for cases `null` and `undefined` . 
It'd be hard for developers to realize the explicit check for `undefined` is needed.


The following is another example of tricky behavior that can stump developers.
```ts
interface Listener {
  (id :number) :void;
}

function playGame(listener? :Listener) :Promise<string> {
  listener = listener  || function() {};

  return new Promise(function(resolve, reject) {
    let id = 9;
    listener(id);  // tsc reports error TS2532: Object is possibly 'undefined' 

    resolve('some result');
  });
} 
```

`tsc` reports that `listener` is possibly `undefined`. From a developer's perspective, he/she would think that `listener` could not be undefined given that it is supplied with a default value (an empty function). Furthermore, there is no other code / closure that could possibly change `listener` back to `undefined`.

I understand the above error can be remedied by using typescript/es2015 default value, or use a non-optional reference, e.g., `let theListener :listener = listener || function() {}` . My point is that these cases could really stump developers:  "The codes have guarded against null, so why does tsc still complain?"

Thanks.
